File: ios/NetworkManager.swift
import Foundation
import Underdark

@objc(NetworkManager)
public class NetworkManager: NetworkCommunicator, ReactNearby {
  fileprivate var type: User.PeerType = User.PeerType.OFFLINE

  //  MARK: REACT NEARBY UTILITY PROTOCOOL
  @objc public  func advertise(_ kind: String) -> Void {
    if self.type == .BROWSER {
      self.type = .ADVERTISER_BROWSER
      self.initTransport("WIFI-BT", inType: .ADVERTISER_BROWSER)
    } else if self.type == .OFFLINE {
      self.type = .ADVERTISER
      self.initTransport("WIFI-BT", inType: .ADVERTISER)
    }
  }
  
  @objc public  func stopAdvertising() {
    if self.type == .ADVERTISER_BROWSER {
      self.type = .BROWSER
      self.initTransport("WIFI-BT", inType: .BROWSER)
      return
    }
    self.type = .OFFLINE
    self.stopTransport()
  }
  @objc public  func browse(_ kind: String) -> Void {
    if self.type == .ADVERTISER {
      self.type = .ADVERTISER_BROWSER
      self.initTransport("WIFI-BT", inType: .ADVERTISER_BROWSER)
    } else if self.type == .OFFLINE {
      self.type = .BROWSER
      self.initTransport("WIFI-BT", inType: .BROWSER)
    }
  }
  
  @objc public  func stopBrowsing() {
    if self.type == .ADVERTISER_BROWSER {
      self.type = .ADVERTISER
      self.initTransport("WIFI-BT", inType: .ADVERTISER)
      return
    }
    self.type = .OFFLINE
    self.stopTransport()
  }
  @objc public  func getConnectedPeers(_ callback: RCTResponseSenderBlock) {
    var connectedPeers = [[String: AnyObject]]()
    if self.type == .BROWSER || self.type == .ADVERTISER_BROWSER {
      for user in nearbyUsers {
        if user.connected {
          connectedPeers.append(user.getJSUser(nil))
        }
      }
    }
    callback([connectedPeers])
  }
  @objc public  func getNearbyPeers(_ callback: RCTResponseSenderBlock) {
    var jsUsers = [[String: AnyObject]]()
    for user in nearbyUsers {
      jsUsers.append(user.getJSUser(nil))
    }
    callback([jsUsers])
  }
  @objc open func inviteUser(_ id: String) {
    let user = self.findUser(id)
    if(user != nil) {
      user!.connected = true
      self.inviteUser(user: user!)
    }
  }
  @objc public func acceptInvitation(_ userId: String) {
    let user = self.findUser(userId)
    if(user != nil) {
      user!.connected = true
      self.informAcceptedInvite(user: user!)
    }
  }
  @objc public  func disconnectFromPeer(_ peerId: String) {
    let user = self.findUser(peerId)
    if(user != nil) {
      user!.connected = false
      self.informDisonnected(user: user!)
    }
  }
  @objc public  func sendMessage(_ message: String, userId:String) {
    self.sendMessage(message: message, userId: userId)
  }
  
  override public func supportedEvents() -> [String]! {
    return ["lostUser","detectedUser", "messageReceived", "connectedToUser", "receivedInvitation"]
  }
}

File: android/src/main/java/com/rctunderdark/NetworkManager.java
package com.rctunderdark;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableArray;

import java.util.Vector;

public class NetworkManager extends ReactContextBaseJavaModule implements ReactNearbyInterface {
    // MARK: private variables
    private User.PeerType type = User.PeerType.OFFLINE;
    private  NetworkCommunicator networkCommunicator;

    // MARK: ReactContextBaseJavaModule
    public NetworkManager(ReactApplicationContext reactContext) {
        super(reactContext);
        this.networkCommunicator = new NetworkCommunicator(reactContext);
    }
    @Override
    public String getName() {
        return "NetworkManager";
    }

    // MARK: React Nearby Utility Methods
    @ReactMethod
    public void sendMessage(String message, String id) {
        this.networkCommunicator.sendMessage(message, id);
    }
    @ReactMethod
    public void advertise(String kind) {
        if(this.type == User.PeerType.BROWSER) {
            this.type = User.PeerType.ADVERTISER_BROWSER;
        } else {
            this.type = User.PeerType.ADVERISER;
        }
        this.networkCommunicator.initTransport(kind, this.type);
    }
    @ReactMethod
    public void browse(String kind) {
        if(this.type == User.PeerType.ADVERISER) {
            this.type = User.PeerType.ADVERTISER_BROWSER;
        } else {
            this.type = User.PeerType.BROWSER;
        }
        this.networkCommunicator.initTransport(kind, this.type);
    }

    @ReactMethod
    public void stopAdvertising() {
        if(this.type == User.PeerType.ADVERTISER_BROWSER) {
            this.type = User.PeerType.BROWSER;
            this.networkCommunicator.initTransport("WIFI-BT", this.type);
            return;
        }
        this.type = User.PeerType.OFFLINE;
        this.networkCommunicator.stopTransport();
    }
    @ReactMethod
    public void stopBrowsing() {
        if(this.type == User.PeerType.ADVERTISER_BROWSER) {
            this.type = User.PeerType.ADVERISER;
            this.networkCommunicator.initTransport("WIFI-BT", this.type);
            return;
        }
        this.type = User.PeerType.OFFLINE;
        this.networkCommunicator.stopTransport();
    }
    @ReactMethod
    public void getConnectedPeers(Callback successCallback) {
        Vector<User> connectedPeers = new Vector<User>();
        Vector<User> nearbyUsers = networkCommunicator.nearbyUsers;
        for(int i = 0; i < nearbyUsers.size(); ++i) {
            if(nearbyUsers.get(i).connected) {
                connectedPeers.add(nearbyUsers.get(i));
            }
        }
        WritableArray jsArray = Arguments.createArray();
        for(int i = 0; i < connectedPeers.size(); ++i) {
            jsArray.pushMap(connectedPeers.elementAt(i).getJSUser());
        }
        successCallback.invoke(jsArray);
    }
    @ReactMethod
    public void getNearbyPeers(Callback successCallback) {
        Vector<User> nearbyUsers = networkCommunicator.nearbyUsers;
        if(nearbyUsers == null) {
            return;
        }
        WritableArray jsArray = Arguments.createArray();
        for(int i = 0; i < nearbyUsers.size(); ++i) {
            jsArray.pushMap(nearbyUsers.elementAt(i).getJSUser());
        }
        successCallback.invoke(jsArray);
    }
    @ReactMethod
    public void inviteUser(String userId) {
        sendMessage("invitation", userId);
    }

    @ReactMethod
    public void acceptInvitation(String userId) {
        User user = networkCommunicator.findUser(userId);
        if(user != null) {
            networkCommunicator.informAcceptedInvite(user);
        }
    }
    @ReactMethod
    public void disconnectFromPeer(String userId) {
        User user = networkCommunicator.findUser(userId);
        if(user != null) {
            user.connected = false;
            networkCommunicator.informDisconnected(user);
        }
    }
}
